### ISSUE TITLE
Allow Layouts in WebService Target

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#1905]
**Type** (choose one):

- Feature request

**NLog version**: 4.4.1

Allow defining layout for webservice. In this case user will have full control of message which will be send to remote server. For example JsonLayout can be used.

```xml
<nlog>
    <target name="logs-service" xsi:type="AsyncWrapper">
      <target xsi:type="WebService" url="http://localhost:5002" protocol="JsonPost">
        <layout xsi:type="JsonLayout">
          <attribute name='logger' layout='${logger}' />
          <attribute name='level' layout='${level}'/>
          <attribute name="message" layout="${message}" />
          <attribute name='parameters' layout='${custom-parameters}' encode="false"/>
         </layout>
      </target>
    </target>
  <rules>
  </rules>
</nlog>
```

as suggested at https://github.com/NLog/NLog/issues/1900

**Comment by yotamLP** [comment]
Hi can you please give me an example on how to implement this custom parameters. I have a web API which I want to send a single json variable. 

do I need to override the webservice method?

**Comment by 304NotModified** [comment]
Do you mean how to implement this feature?

**Comment by yotamLP** [comment]
yes

**Comment by snakefoot** [comment]
The WebService-target needs to have a parameter to provide the payload, because it supports both HttpGet and HttpPost.

Just like there is XmlRoot-parameter to wrap all parameters inside a single XML-tag, then one could create a new bool parameter called SingleParameterRoot. That promotes the first parameter to be root when doing HttpPost. Whether it is Json or Xml. When SingleParameterRoot=true then it will validate that only a single parameter has been configured.

**Comment by snakefoot** [comment]
Could also consider leaving the WebService-target behind, and instead create a new target based on HttpClient (Net45-only). Then one doesn't have to fight with the backward compatibility dragons and can use recommended high performance APIs.

**Comment by ParallelTask** [comment]
One of my projects requires the exact thing, so I have created NLog.RestTarget that solves the exact problem. Please find the reference https://github.com/ParallelTask/Nlog.RestTarget

**Comment by snakefoot** [comment]
There is now a build that supports this:

https://www.nuget.org/packages/NLog/4.5.0-rc07

Notice the use of a single nameless parameter:

```xml
<target xsi:type="WebService" url="http://localhost:5002" protocol="JsonPost">
  <parameter name="">
        <layout xsi:type="JsonLayout">
          <attribute name='logger' layout='${logger}' />
          <attribute name='level' layout='${level}'/>
          <attribute name="message" layout="${message}" />
          <attribute name='parameters' layout='${custom-parameters}' encode="false"/>
         </layout>
  </parameter>
</target>
```

**Comment by lemunk** [comment]
I see the new build supports it but how would we use this with our custom parameters?

**Comment by snakefoot** [comment]
Can you give an example of what you mean, when saying custom parameters ?


Sent from my Samsung device


-------- Original message --------

**Comment by snakefoot** [comment]
Resolved by #2590
